buildSrc/src/main/java/org/elasticsearch/gradle/test/ErrorReportingTestListener.java

```diff
@@ -0,0 +1,180 @@
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class ErrorReportingTestListener implements TestOutputListener, TestListener {
+    private static final String REPRODUCE_WITH_PREFIX = "REPRODUCE WITH";
+
+    private final TestExceptionFormatter formatter;
+    private Map<Descriptor, List<TestOutputEvent>> eventBuffer = new ConcurrentHashMap<>();
+    private Set<Descriptor> failedTests = new LinkedHashSet<>();
+
+    public ErrorReportingTestListener(TestLogging testLogging) {
+        this.formatter = new FullExceptionFormatter(testLogging);
+    }
+
+    @Override
+    public void onOutput(TestDescriptor testDescriptor, TestOutputEvent outputEvent) {
+        TestDescriptor suite = testDescriptor.getParent();
+
+        // Check if this is output from the test suite itself (e.g. afterTest or beforeTest)
+        if (testDescriptor.isComposite()) {
+            suite = testDescriptor;
+        }
+
+        List<TestOutputEvent> events = eventBuffer.computeIfAbsent(Descriptor.of(suite), d -> new ArrayList<>());
+        events.add(outputEvent);
+    }
+
+    @Override
+    public void beforeSuite(TestDescriptor suite) {
+
+    }
+
+    @Override
+    public void afterSuite(final TestDescriptor suite, TestResult result) {
+        try {
+            // if the test suite failed, report all captured output
+            if (result.getResultType().equals(TestResult.ResultType.FAILURE)) {
+                List<TestOutputEvent> events = eventBuffer.get(Descriptor.of(suite));
+
+                if (events != null) {
+                    // It's not explicit what the threading guarantees are for TestListener method execution so we'll
+                    // be explicitly safe here to avoid interleaving output from multiple test suites
+                    synchronized (this) {
+                        System.err.println("\n\nSuite: " + suite);
+
+                        for (TestOutputEvent event : events) {
+                            log(event.getMessage(), event.getDestination());
+                        }
+                    }
+                }
+            }
+        } finally {
+            // make sure we don't hold on to test output in memory after the suite has finished
+            eventBuffer.remove(Descriptor.of(suite));
+        }
+    }
+
+    @Override
+    public void beforeTest(TestDescriptor testDescriptor) {
+
+    }
+
+    @Override
+    public void afterTest(TestDescriptor testDescriptor, TestResult result) {
+        if (result.getResultType() == TestResult.ResultType.FAILURE) {
+            failedTests.add(Descriptor.of(testDescriptor));
+
+            if (testDescriptor.getParent() != null) {
+                // go back and find the reproduction line for this test failure
+                List<TestOutputEvent> events = eventBuffer.get(Descriptor.of(testDescriptor.getParent()));
+                for (int i = events.size() - 1; i >= 0; i--) {
```

Comment:
This is a matter of personal preference. 
I find streams more readable. This could be `IntStream.map().filter().foreach(System.err::print)`

Reply:
That's not the same here as we iterating _backward_ through this list and then stopping when we find the first matching candidate.

Reply:
You could handle that from the map as the size is known, but I don't have a strong preference.